src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -254,4 +267,88 @@ void justClose() {
             SCOPED_MEMORY_ACCESS.closeScope(this);
         }
     }
+
+    /**
+     * A functional interface modelling the cleanup action associated with a scope.
+     */
+    interface CleanupAction {
+        void cleanup();
+        CleanupAction dup();
+
+        /** Dummy cleanup action */
+        CleanupAction DUMMY = new CleanupAction() {
+            @Override
+            public void cleanup() {
+                // do nothing
+            }
+
+            @Override
+            public CleanupAction dup() {
+                return this;
+            }
+        };
+    }
+
+    /**
+     * A stateful cleanup action; this action can only be called at most once. The implementation
+     * guarantees this invariant even when multiple threads race to call the {@link #cleanup()} method.
+     */
+    static abstract class BasicCleanupAction implements CleanupAction {
```

Comment:
Rename to `AtMostOnceCleanupAction` or some other snappier name that conveys it's purpose. Embedded in `CleanupAction` then `CleanupAction.AtMostOnceOnly` might work.